Rewrite OpenQASM 3 exporter symbol table

jakelishman opened the pull request:
### Summary

This rewrites the symbol handling of the OpenQASM 3 exporter to decouple object identities from the necessary object identifiers.  As part of this, we use the same trick of standard-gate reparametrisation to produce gate-definition sources for Qiskit built-ins, which fixes many cases of bad parametrisation of gates like `rzx`.

This kind of rewrite was necessary to fix the now-bad assumption within the OQ3 exporter that "gate identity" is always static within a circuit. Since standard gate `Gate` instances are now only generated on demand, there is no guarantee of stability of them.  The fix to the definition source for these makes them independent of object identity. User-defined gates can still use the identity, as these are still guaranteed static.

This commit fixes almost all of the "bad parametrisation" tests in the test suite.  There are several other changes in the test suite necessary:

* since the uniqueness of the identifier is now independent of how the lookup of a Qiskit object works, there is no need to include the highly non-deterministic `id` in the generated symbols for user gates. Several tests changed to use the new, simple count-based unique names.

* the escaping and uniqueness rules now apply uniformly to all gate definitions, fixing several bad test cases that previously were testing invalid OpenQASM 3.

* the escaping rules changed slightly for naming collisions with keywords, making them slightly more consistent with how other renaming rules worked.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Details and comments

This unblocks #12730 by fixing the broken assumption about object identity for standard gates.

* Fix #7136 (not to say it's perfect now, but this addresses all the problems tracked by that issue).
* Fix #7335
* Fix #9428
* Fix #11954
* Fix #11560
* Fix #11761

* Supersede #12481
    This fixes the underlying bug that PR 12481 was attempting to address, though reducing the use of `instruction.operation` as that PR does could still be a benefit to performance.

jakelishman committed:
Rewrite OpenQASM 3 exporter symbol table

This rewrites the symbol handling of the OpenQASM 3 exporter to decouple
object identities from the necessary object identifiers.  As part of
this, we use the same trick of standard-gate reparametrisation to
produce gate-definition sources for Qiskit built-ins, which fixes many
cases of bad parametrisation of gates like `rzx`.

This kind of rewrite was necessary to fix the now-bad assumption within
the OQ3 exporter that "gate identity" is always static within a circuit.
Since standard gate `Gate` instances are now only generated on demand,
there is no guarantee of stability of them.  The fix to the definition
source for these makes them independent of object identity.
User-defined gates can still use the identity, as these are still
guaranteed static.

This commit fixes almost all of the "bad parametrisation" tests in the
test suite.  There are several other changes in the test suite
necessary:

* since the uniqueness of the identifier is now independent of how the
  lookup of a Qiskit object works, there is no need to include the
  highly non-deterministic `id` in the generated symbols for user gates.
  Several tests changed to use the new, simple count-based unique names.

* the escaping and uniqueness rules now apply uniformly to all gate
  definitions, fixing several bad test cases that previously were
  testing invalid OpenQASM 3.

* the escaping rules changed slightly for naming collisions with
  keywords, making them slightly more consistent with how other renaming
  rules worked.

jakelishman commented:
Marking as high priority because this blocks #12730.

jakelishman commented:
The failing test is fixed by #12775.  I'll push a fix to the lint failure once that's merged.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -115,13 +120,8 @@
-    if name in _RESERVED_KEYWORDS or not _VALID_IDENTIFIER.fullmatch(name):
-        name = "_" + re.sub(r"[^\w\d]", "_", name)
-    return name
+_VALID_IDENTIFIER = re.compile(r"(\$[\d]+|[\w][\w\d]*)", flags=re.U)
```
EDIT: This is fixed by https://github.com/jakelishman/qiskit-terra/pull/25

This allows things like `register_gate_without_definition("$1", None)`. ~~Is this intended?~~

For example
```python
mygate = Gate("$1", 1, [])
qc = QuantumCircuit(1)
qc.append(mygate, [0])
qcdef = QuantumCircuit(1)
qcdef.x(0)
mygate.definition = qcdef
```
writes an OQ3 gate named `$1`  when exported, whereas `Gate("&1"...` is properly escaped.

jlapeyre replied:
I might factor out the two uses of this regular expression with
```python
def is_valid_identifier(string):
    return _VALID_IDENTIFIER.fullmatch(string)
```
But it's ok the way it is, too.

A comment: Because of Perl habits, I would have used `^` and `$` to anchor the pattern instead of `fullmatch`. But I think using `fullmatch`  (regardless of whether the matching is refactored)  makes it easier to understand the code, both because the reader doesn't have to look back to the regular expression to know if the entire string must match, and becuase they need less knowledge of regular expressions.

jakelishman replied:
I cherry-picked jakelishman/qiskit-terra#25 (with minor modifications) as 598fae3b5a2df29a9aeba4d8bac6bdc3f7656d89.

imo the ^$ anchoring restricts the regex beyond what its name (`_VALID_IDENTIFIER`) actually suggests - it works well when the regex is defined inline, but I agree that `fullmatch` is clearer when it's encapsulated in a variable.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -829,10 +900,13 @@ def build_current_scope(self) -> List[ast.Statement]:
+            for var in self.scope.circuit.iter_declared_vars()
         ]
-        for instruction in scope.circuit.data:
+        for instruction in self.scope.circuit.data:
```
As mentioned in the PR description, a possible remaining benefit
of the superseded #12481 is fewer attribute lookups: `instruction.operation`.

However, the code block that's touched in #12481 is removed in the present PR, #12776.
I guess the time to make attribute lookups is negligible compared to other tasks, such
as string building (even though this is done by copying short strings to a stream).

A remaining benefit in my view is readability. But that's debatable. You might consider factoring out
the attribute lookup here. It's fine as it is, as well.

jakelishman replied:
My plan was to leave this for your PR anyway - I wasn't trying to override that.

jlapeyre replied:
Yes, thanks for that. We can think about that after this PR then.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+_FIXED_PARAMETERS = (Parameter("p0"), Parameter("p1"), Parameter("p2"), Parameter("p3"))
+
+# Mapping of symbols defined by `stdgates.inc` to their gate definition source.
+_KNOWN_INCLUDES = {
```
I might do something like this. Just a suggestion:
```python
from qiskit.circuit import(Gate, ControlledGate, Parameter, library)

_FIXED_PARAMETERS = (Parameter("p0"), Parameter("p1"), Parameter("p2"), Parameter("p3"))

def _gate_definition_source(gate: Gate) -> Gate:
    if gate._standard_gate and not isinstance(gate, ControlledGate):
        return gate.base_class(*_FIXED_PARAMETERS[: len(gate.params)])
    elif gate._standard_gate:
        return gate.base_class(*_FIXED_PARAMETERS[: len(gate.params)], ctrl_state=gate.ctrl_state)
    return gate

def _std_known_includes():
    std_gate_names = ['p', 'x', 'y', 'z', 'h', 's', 'sdg', 't', 'tdg',
                      'sx', 'rx', 'ry', 'rz', 'cx', 'cy', 'cz', 'cp',
                      'crx', 'cry', 'crz', 'ch', 'swap', 'ccx', 'cswap',
                      'cu', 'id', 'u1', 'u2', 'u3']
    alias_map = {'cx': 'CX', 'p': 'phase', 'cp': 'cphase'}

    qiskit_map = library.get_standard_gate_name_mapping()
    source_gates = {name: _gate_definition_source(qiskit_map[name]) for name in std_gate_names}
    for name, alias in alias_map.items():
        source_gates[alias] = _gate_definition_source(source_gates[name])
    return source_gates

_KNOWN_INCLUDES = {"stdgates.inc": _std_known_includes()}

```

jlapeyre replied:
In fact, I'm looking at the long list of lines like ` CustomGate(library.XGate, "x", 0, 1)` in `__init__.py` and the list here, and the list in `library.get_standard_gate_name_mapping`. It seems like it would not be hard to encapsulate this data in one place with simple API.

The snippet above is a nod in this direction. It reduces the sources of truth a bit.

jakelishman replied:
The mapping between Qiskit's gate names and the name that they're given in OQ2 and OQ3 isn't entirely identical (Qiskit's OQ2 `qelib.inc` extension calls `c3sx` something different, iirc), and there's the aliasing changes, so I wouldn't necessarily say that there's a duplication in sources of truth here; the "linking" between Qiskit's source of truth and OpenQASM's is itself a new set of facts to be encoded.

Personally I wrote it the way I did because I find the (relatively short) table easier to read and understand than generation code, and there's less to debug in terms of logic.  The more code I write, I find that I'm raising my internal barrier to wanting to write "code that writes code" and so on - I used to do it a lot more, then it was often more effort to maintain than a slightly longer but more explicit list.

Obviously there's a limit to that - I'm of course not suggesting that tablegen has no purpose.

jlapeyre replied:
Sure, those are good arguments. Simple, readable, self-contained code is nice to find.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+    node: ast.QuantumGateDefinition | None
+    """An AST node containing the gate definition.  This can be ``None`` if the gate came from an
+    included file, or is an overridden "basis gate" of the export."""
+
```
If I understand correctly, `GateDefinition` links the qiskit "definition" with the OQ3 definition. And the qiskit definition is a `Gate` object instance with canonical parameters (fixed parameters shared by all gates).
The word "definition" is used here and repeatedly throughout the file. It's often not immediatley clear whether we are referring to the qiskit or the OQ3 defnition.  I wonder if we can find terms that make it easier to disitinguish.

For example, here we have "definition source of the gate" and "gate definition".  The word "source" is only used in this file for qiskit representations. So that must mean the second "gate definition" refers to OQ3. And of course `Gate` is a qiskit class representing gates. And `ast` must mean qasm. 

Maybe "qiskit definition source" and "OpenQASM 3 gate defintion". And also maybe state explicilty that this is the data that associates two representations in the two systems.

jakelishman replied:
Good idea, thanks.  Much of what we're doing is object canonicalisation for equality comparisons, so I've renamed a bunch of things around that idea in 771676c412.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+        If the ``gate`` is given, it will be used to validate that a call to it is compatible (such
+        as a known gate from an included file).  If it is not given, it is treated as a user-defined
+        "basis gate" that assumes that all calling signatures are valid and that all gates of this
+        name are exactly compatible, which is somewhat dangerous."""
```
"somewhat dangerous"  haha

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+        else:
+            source = _gate_definition_source(gate)
+            self.gates[name] = GateDefinition(source, None)
+            if (standard_gate := getattr(source, "_standard_gate")) is not None:
```
Why is it `getattr` here, but `._standard_gate` in `_gate_definition_source`?

jakelishman replied:
I'm not entirely sure.  Everything's required to be `Gate` (or `None`, but that's already handled) throughout these paths, so the attribute always exists.  Removed unnecessary `getattr` in b51386fac2.

jlapeyre reviewed (commented):
This looks great so far. I made a few comments.

I still have a couple of questions on the design. When I finish reading exporter.py tomorrow they may evaporate.
EDIT: The questions dissipated.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -502,172 +648,98 @@ def build_program(self):
+        This recurses through gate-definition statements."""
+        if issubclass(gate.base_class, library.CXGate) and gate.ctrl_state == 1:
+            # CX gets super duper special treatment because it's the base of Qiskit's definition
+            # tree, but isn't an OQ3 built-in (it was in OQ2).  We use `isinstance` because we
```
We use `issubclass` ?

jakelishman replied:
We don't subclass `CXGate` within Qiskit, but I think think this is what would be expected if we did.  It also avoids pylint complaining about unidiomatic type checks (`gate.base_class is library.CXGate` would be a complaint, I think).

jlapeyre replied:
My comment was not clear!  I meant that the doc string says "We use `isinstance`", the code uses `issubclass`

jakelishman replied:
Oh I see it now - thanks!  Fixed in 984c4220c1.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+
+        This is insertion ordered, and that can be relied on for iteration later."""
+        self.standard_gate_idents: dict[StandardGate, ast.Identifier] = {}
+        """Mapping of standard gate enumeration values to the identifier we represent that as."""
+        self.user_gate_idents: dict[int, ast.Identifier] = {}
+        """Mapping of `id`s of user gates to the identifier we use for it."""
+
+        self.variables: list[dict[str, object]] = [{}]
+        """Stack of mappings of variable names to the Qiskit object that represents them.
+
+        The zeroth index corresponds to the global scope, the highest index to the current scope."""
+        self.objects: list[dict[object, ast.Identifier]] = [{}]
+        """Stack of mappings of Qiskit objects to the identifier (or subscripted identifier) that
+        refers to them.  This is similar to the inverse mapping of ``variables``.
```
(See also the previous comment on descriptions of mappings)
I'm looking for a way to describe a map without too much ambiguity and subject/verb disagreement. Maybe something like:
```suggestion
        """Mapping from standard gate enumeration values to the identifiers we represent them as."""
        self.user_gate_idents: dict[int, ast.Identifier] = {}
        """Mapping from `id`s of user gates to the identifiers we use for them."""

        self.variables: list[dict[str, object]] = [{}]
        """Stack of mappings from variable names to the Qiskit objects that represent them.

        The zeroth index corresponds to the global scope, the highest index to the current scope."""
        self.objects: list[dict[object, ast.Identifier]] = [{}]
        """Stack of mappings from Qiskit objects to the identifiers (or subscripted identifiers) that
        refer to them.  This is similar to the inverse mapping of ``variables``.
```

jakelishman replied:
Leaving as is, per https://github.com/Qiskit/qiskit/pull/12776#issuecomment-2242631327.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+# comparisons will work.
+_FIXED_PARAMETERS = (Parameter("p0"), Parameter("p1"), Parameter("p2"), Parameter("p3"))
+
+# Mapping of symbols defined by `stdgates.inc` to their gate definition source.
```
I think I actually read this smoothly without big hitches for disambiguation. But it still might be more clear like this:
```suggestion
# Mapping from symbols defined by `stdgates.inc` to their gate definition sources.
```

jakelishman replied:
Leaving as is, per https://github.com/Qiskit/qiskit/pull/12776#issuecomment-2242631327.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+
+    def __init__(self):
+        self.gates: collections.OrderedDict[str, GateDefinition | None] = {}
+        """Mapping of the symbol name to the "definition source" of the gate, which provides its
```
Elsewhere, I make suggested changes to the descriptions of mappings. But line 264 is a good  choice.

Two good ways to describe the maps are

"Mapping of the red item to the blue item"

"Mapping from the red items to the blue items"

jakelishman replied:
I've just "mapping of" consistently, and I'm not sure I agree there's any distinction in the preposition implied on plurality - both read fine to me.

jakelishman replied:
Leaving as is, per https://github.com/Qiskit/qiskit/pull/12776#issuecomment-2242631327.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+    circuit: QuantumCircuit
+    """The circuit block that we're currently working on exporting."""
+    bit_map: dict[Bit, Bit]
+    """Mapping of bit objects in ``circuit`` to the bit objects in the global-scope program
```
This is readable. I don't think the English-language description needs to be mathematically precise, as long as a reader easily and unconsciously gets a precise understanding.
I'd suggest making the pattern of descriptions uniform, or at least intentionally choosing the pattern based on readability.

In most cases, this is a very minor point. But it also would take only a couple of minutes to edit the 10 or so cases.
It's somewhat more important to make a good choice with more complicated objects, such  as "stack of mappings".

jakelishman replied:
What are you thinking isn't precise about this statement?

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+        self.objects.pop()
+        self.variables.pop()
+
+    def new_context(self) -> SymbolTable:
```
You might consider `def new_symbol_table(self) -> SymbolTable:`. This is a method in the class `SymbolTable` and the doc strings elsewhere in the class refer to `SymbolTable`. I suppose developer docs will never be extracted and formatted. But if they were, the method name and description here would be puzzling.

I find the contexts, symbol tables, and scopes a bit confusing. Maybe it's unavoidable. But a brief overview might be helpful.

jakelishman replied:
I think the docstring of this method fairly clearly defines what a "new context" is?  imo, `new_symbol_table` erases part of the description of the method - if I read it, I'd think "what's the difference from `SymbolTable()`?".

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+
+
+@dataclasses.dataclass
+class BuildScope:
```
There are two kinds of scope, right? Is it: The scope in `SymbolTable` is an OQ3 scope. And `BuildScope` is a scope relevant only to the process of exporting... or maybe a qiskit-side scope. In the doc string, we have
"top-level scope where the bits were actually defined". Does "scope" here refer to qiskit? Maybe using terms "qiskit-scope" and "qasm-scope", or whatever the correct categories are, would be helpful.

jakelishman replied:
The "scope" in `SymbolTable` and `BuildScope` refer to the same concept - I was trying to unify those names.  OpenQASM 3 has more than one type of scope (gates introduce different scoping rules to non-gate blocks, e.g.), which is also represented here by the idea of a "context", but a context still includes a current scope.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+    """Track Qiskit objects and the OQ3 identifiers used to refer to them."""
+
+    def __init__(self):
+        self.gates: collections.OrderedDict[str, GateDefinition | None] = {}
```
Did you consider keeping everything in the same stack of symbol tables? I suppose you'd have attach a type or other label to each object, in order to do this. But I don't see a lot of types otherwise. To me, keeping the gates separate like you have done looks cleaner and more manageable. Maybe it's the only reasonable way.

jakelishman replied:
Gates have different rules to variables with regards to definitions, shadowing, etc, so it's a bit clearer (imo) to track them separately - for example, there's no concept of scopes for gate identifiers, since they've got to be global definitions and can't be shadowed ever.

We also need to track different information for `Gate` objects compared to other symbols, so it's convenient to put them in a separate table.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+        return self.user_gate_idents.get(id(source))
+
+
+def _gate_definition_source(gate: Gate) -> Gate:
```
I've been wondering why you don't store something like `(GateClass, num_params)`, instead of constructing a gate with fixed, otherwise useless, params. It looks like there are several reasons. One is that we don't have a convenient interface for obtaining information about a gate other than calling methods on instances of the gate. So, you'd have to construct a gate anyway to get it. Another is you don't try to do anything clever with user gates; they are stored as instances, and are their own source, and you hope for the best. So if you treated standard gates differently, your definition of source would have a more heterogeneous format. Although this is a sort of vague view of the problem, trying to store the definition source differently seems to cause problems, and not to really solve any.

What you have here is pretty straightforward, given the constraints of current qiskit data models.

jakelishman replied:
Hopefully this is a little clearer with the renaming of this method to `_gate_canonical_form`.  It's clearer for typing and usage if everything's always just a `Gate` (or `None`) - no need to do type reflection on lookups, just use the same comparisons as always.

jlapeyre commented on qiskit/qasm3/exporter.py:
```diff
@@ -364,108 +531,86 @@ def __init__(
-        gate_definitions = [
-            self.build_gate_definition(instruction) for instruction in gates_to_declare
-        ]
+        # The order we build parts of the AST has an effect on which names will get escaped to avoid
+        # collisions.  The current ideas are:
```
This is pretty clear and useful.

jlapeyre commented on test/python/qasm3/test_export.py:
```diff
@@ -494,7 +542,7 @@ def test_unbound_circuit(self):
         self.assertEqual(Exporter().dumps(qc), expected_qasm)
 
-    def test_unknown_parameterized_gate_called_multiple_times(self):
+    def test_standard_parameterized_gate_called_multiple_times(self):
```
EDIT: I pushed a commit add a test here https://github.com/jakelishman/qiskit-terra/pull/24

I don't see a test for the case that first instance of the gate is **not** generic.
```python
        qc.rzx(0.5, 0, 1)
        qc.rzx(x, 0, 1)
        qc.rzx(y, 0, 1)
```
With this PR this is exported correctly!

jakelishman replied:
Thanks - cherry-picked as 2cbf15ff4d0ae35e845e9db9a66070597b14b679.

jlapeyre reviewed (commented):
@jakelishman  This is the last part of my review. 

Above, I mentioned a design question. I wondered why a choice was made. I think I see why and mention this in a comment.

This looks like a great improvement. I was not able to go over all of it.

One of the more important comments below is an apparently missing test for a bug fix. (The test still assumes the bug!)

jlapeyre committed:
Add test for bug fix for issue #7335

jlapeyre committed:
Rename qiskit gates whose names are OQ3 hardware qubit identifiers

If a custom qiskit gate is given a name that is a valid identifer for a
hardware qubit in OQ3, then, before this commit, the name would not be
escaped when writing the OQ3 gate definition.

This commit fixes this by escaping the leading dollar sign as it would
be in any other position in the name. That is, the dollar sign is
replaced by underscore.

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

jlapeyre commented:
TLDR; https://github.com/jakelishman/qiskit-terra/pull/25 and https://github.com/jakelishman/qiskit-terra/pull/24 fix a couple of oversights/bugs in this PR.

jakelishman committed:
Reduce overloading of word "definition"

Much of what we're doing with the "definition source" is actually a form
of object canonicalisation for comparison purposes.  It's clearer to use
this terminology.

jakelishman committed:
Remove unnecessary getattr

jakelishman commented:
I've addressed all of the above actionable comments.  I left the preposition/plurality in "mapping of X(s) to Y(s)" as-is, since I don't think there's any meaningful difference, and we already use both "mapping of" and "mapping from" with plurals and singulars throughout Qiskit.

jakelishman commented:
The last couple of remaining comments are about potentially improving internal developer documentation in the exporter.  That could well be valuable for the future, but let's do it in a follow-up if so, so we can get this merged and unblock #12730 in the interim.

jakelishman committed:
Fix isinstance/issubclass naming

jlapeyre commented:
If you've reviewed the language around mappings and are satisfied, then so am I.

jlapeyre reviewed (approved):
Looks great

ElePT commented on test/python/qasm3/test_export.py:
```diff
@@ -429,11 +422,14 @@ def test_custom_gate_with_params_bound_main_call(self):
 
         circuit.assign_parameters({parameter0: pi, parameter1: pi / 2}, inplace=True)
 
+        # NOTE: this isn't exactly what we want; note that the parameters in the signature are not
+        # actually used.  It would be fine to change the output of the exporter to make `custom` non
+        # parametric in this case.
```
(just to make sure I am following) This isn't what we want because we want the parameters in the signature to be used in the definition instead of `pi`/`pi/4`? Where would the parameter value be stored then?

I can see that at least the issue from #7335 (the value being copied over) isn't there anymore.

jakelishman replied:
Ideally, the output of this test should be
```qasm
OPENQASM 3.0;
include "stdgates.inc";
qubit[2] q;
gate custom(p0, p1) q0, q1 {
  rz(p0) q0;
  rz(p1 / 2) q1;
}
custom(pi, pi/2) q[0], q[1];
```
since that's how the `custom` gate is originally defined in Qiskit.  We can't actually do that, though, because the prior call to `assign_parameters` erases the original parametric definition, so we can't find it again.  The test as written doesn't use `p0` or `p1`.

With this PR, this "bad" behaviour is only limited to custom user gates, though, so it's not such a huge deal anymore, as it's rather less likely to come up in practice.

ElePT reviewed (approved):
LGTM too!
